Index: linux-3.10.49/drivers/net/ethernet/ralink/ralink_soc_eth.c
===================================================================
--- linux-3.10.49/drivers/net/ethernet/ralink/ralink_soc_eth.c
+++ linux-3.10.49/drivers/net/ethernet/ralink/ralink_soc_eth.c
@@ -1489,9 +1489,9 @@
 
 	napi_weight = 32;
 	if (priv->flags & FE_FLAG_NAPI_WEIGHT) {
-		napi_weight *= 2;
-		priv->tx_ring_size *= 2;
-		priv->rx_ring_size *= 2;
+		napi_weight *= 4;
+		priv->tx_ring_size *= 4;
+		priv->rx_ring_size *= 4;
 	}
 	netif_napi_add(netdev, &priv->rx_napi, fe_poll, napi_weight);
 	fe_set_ethtool_ops(netdev);
