Index: linux-3.10.49/include/uapi/linux/usb/ch11.h
===================================================================
--- linux-3.10.49/include/uapi/linux/usb/ch11.h
+++ linux-3.10.49/include/uapi/linux/usb/ch11.h
@@ -11,6 +11,17 @@
 
 #include <linux/types.h>	/* __u8 etc */
 
+/* This is arbitrary.
+ * From USB 2.0 spec Table 11-13, offset 7, a hub can
+ * have up to 255 ports. The most yet reported is 10.
+ *
+ * Current Wireless USB host hardware (Intel i1480 for example) allows
+ * up to 22 devices to connect. Upcoming hardware might raise that
+ * limit. Because the arrays need to add a bit for hub status data, we
+ * use 31, so plus one evens out to four bytes.
+ */
+#define USB_MAXCHILDREN		31
+
 /*
  * Hub request types
  */
Index: linux-3.10.49/include/uapi/linux/usb/functionfs.h
===================================================================
--- linux-3.10.49/include/uapi/linux/usb/functionfs.h
+++ linux-3.10.49/include/uapi/linux/usb/functionfs.h
@@ -32,12 +32,13 @@
  * All numbers must be in little endian order.
  */
 
+/* Legacy format, deprecated as of 3.14. */
 struct usb_functionfs_descs_head {
 	__le32 magic;
 	__le32 length;
 	__le32 fs_count;
 	__le32 hs_count;
-} __attribute__((packed));
+} __attribute__((packed, deprecated));
 
 /*
  * Descriptors format:
